File: leetcode/medium/221-maximal-square/iter-dp.py
"""
  @user Vortexx2
  Problem 221 - Maximal Square

  Runtime - 586 ms
  Memory Usage - 16.4 MB
"""
from typing import List


class Solution:
  def maximalSquare(self, matrix: List[List[str]]) -> int:

    m, n = len(matrix), len(matrix[0])

    self.maxSide = 0

    """
      Returns the bottom most row for the DP that is required
    """
    def bottomMostRow() -> List[int]:
      res: List[int] = [1] * n

      for i, num in enumerate(matrix[m - 1]):

        if num == "0":
          res[i] = 0

        else:
          self.maxSide = 1
          res[i] = 1

      return res

    # prevRow is the DP that maps the maximal square taking that position as the top left part of the square
    prevRow = bottomMostRow()

    for i in range(m - 2, -1, -1):
      currRow = [1] * n

      for j in range(n - 1, -1, -1):

        # if character is 0, maximal square possible is 0
        if matrix[i][j] == "0":
          currRow[j] = 0

        # if character is 1
        else:

          # if it's not the last position of the column
          if j != n - 1:
            currRow[j] = 1 + min(currRow[j + 1], prevRow[j], prevRow[j + 1])
            self.maxSide = max(self.maxSide, currRow[j])

          # if it's the last position in the column, and character is 1
          else:
            currRow[j] = 1
            self.maxSide = max(self.maxSide, 1)

      prevRow = currRow

    return self.maxSide * self.maxSide

File: leetcode/medium/221-maximal-square/iter-dp.cpp
/**
 * @file iter-dp.cpp
 * @author Vortexx2
 * @brief Problem 221 - Maximal Square
 * @approach Tabulation DP approach with the least memory possible
 * @date 07-08-2022
 *
 * Runtime - 101 ms O(m * n)
 * Memory Usage - 18 MB O(n)
 */
#include <iostream>
#include <vector>

using namespace std;

class Solution {
 public:
  int maxSide = 0;
  int maximalSquare(vector<vector<char>>& matrix) {
    int m = matrix.size(), n = matrix[0].size();

    vector<int> prevRow = bottomRow(matrix);

    for (int i = m - 2; i >= 0; i--) {
      vector<int> currRow(n, 1);
      for (int j = n - 1; j >= 0; j--) {
        if (matrix[i][j] == '0')
          currRow[j] = 0;

        else {
          // if it is not the last column and char is 1
          if (j != n - 1)
            currRow[j] =
                1 + min(currRow[j + 1], min(prevRow[j], prevRow[j + 1]));

          // if it is the last column and char is 1
          else
            currRow[j] = 1;

          maxSide = max(maxSide, currRow[j]);
        }
      }

      prevRow = currRow;
    }

    return maxSide * maxSide;
  }

  vector<int> bottomRow(vector<vector<char>>& matrix) {
    int rows = matrix.size(), cols = matrix[0].size();
    vector<int> res(cols);

    for (int i = 0; i < cols; i++) {
      if (matrix[rows - 1][i] == '0')
        res[i] = 0;

      else {
        maxSide = 1;
        res[i] = 1;
      }
    }

    return res;
  }
};

int main() {
  Solution sol = Solution();
  vector<vector<char>> a = {{'1', '0'}, {'0', '0'}};

  cout << sol.maximalSquare(a) << endl;
}